LNK-2740: Measure Evaluator Unit Tests

c-schuler opened the pull request:
Work done so far:
- Instantiation tests for MeasureEvaluator
- Evaluation tests for cohort, proportion, ratio and continuous variable measures
- Setup builders for patient data and artifacts

TODO:
- [x]  Harden evaluator logging and validation
- [ ]  Plugin a better caching strategy 
- [ ]  Add tests for:
  - [x]  SDEs
  - [x]  Terminology
  - [ ]  Edge cases

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced robust measure evaluation tests for cohort, proportion, ratio, and continuous variable measures.
  - Added the `PatientDataBuilder` for streamlined creation of FHIR resources in tests.

- **Bug Fixes**
  - Enhanced error handling in `MeasureEvaluator` for unsupported FHIR versions and empty bundle entries.

- **Tests**
  - Introduced comprehensive test cases for `MeasureEvaluator` instantiations and evaluations.
  - Added methods to `PatientDataBuilder` for constructing test data.

- **Chores**
  - Added JUnit Jupiter dependency for enhanced testing capabilities.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

c-schuler committed:
Adding unit tests - initial setup

- Instantiation tests for MeasureEvaluator
- Evaluation tests for cohort, proportion, ratio and continuous variable measures
- Setup builders for patient data and artifacts

c-schuler committed:
Evaluator logging and validation hardening

seanmcilvenna commented:
@coderabbitai review

seanmcilvenna commented on Java/measureeval/src/test/java/com/lantanagroup/link/measureeval/services/KnowledgeArtifactBuilder.java:
```diff
@@ -0,0 +1,283 @@
+        public static Measure measure() {
+            Measure measure = new Measure();
+            measure.addLibrary(LIBRARY_URL);
+            measure.setScoring(new CodeableConcept().addCoding(new Coding().setCode("proportion")));
```
We're primarily targeting cohort measures, right now. I like testing additional scoring methods here, but can we make sure that "Cohort" is among the tests, since that's what we're using now?

c-schuler replied:
Added additional cohort measure tests

seanmcilvenna reviewed (changes requested):
* Can we get some tests for other classes in the service, like the `MeasureCache` class?
* Would like to see some negative tests, as well (so-to-speak)... which would require only a _slightly_ more complex CQL that does more than `return true`
* One of the things that is done often in our use case is to have definitions that return full resources (which get represented as contained resources on the resulting MeasureReport. Can you include a test that demonstrates that successfully?

c-schuler committed:
Applying feedback

Additional tests: IP false Cohort, Cohort with ValueSet and evaluated resources, Cohort with SDE and contained

Applied several coderabbitai suggestions: made builders/tests more extensible and cleaned up code

More to come!

smailliwcs commented on Java/measureeval/src/test/java/com/lantanagroup/link/measureeval/services/MeasureEvaluatorInstantiationTests.java:
Agree with the intent behind these test cases, but they should probably be applied to `MeasureDefinitionBundleValidator`, because that's the class we use to validate client-provided bundles.  By the time we get to `MeasureEvaluator` instantiation (which is what we're testing here), it's too late to reject an invalid bundle.

c-schuler replied:
@smailliwcs I have applied your feedback. Thanks!

c-schuler committed:
Merge branch 'dev' into me-unit-tests

c-schuler committed:
Merge branch 'dev' into me-unit-tests

kissalk committed:
Update Dockerfile

c-schuler committed:
Added measure package validation - cleaned up builders and CQL

c-schuler committed:
Merge branch 'dev' into me-unit-tests

c-schuler committed:
Added more complex false test

c-schuler committed:
Merge branch 'dev' into me-unit-tests

c-schuler committed:
Added JavaDoc

c-schuler committed:
Merge branch 'dev' into me-unit-tests

seanmcilvenna commented:
@c-schuler can you make sure this branch gets updated with the latest from `dev`?
@smailliwcs can you give look over this?

c-schuler committed:
Merge branch 'dev' into me-unit-tests

c-schuler commented on Java/measureeval/src/test/java/com/lantanagroup/link/measureeval/services/PatientDataBuilder.java:
```diff
@@ -0,0 +1,34 @@
+        var patient = new Patient();
+        patient.setId("Patient/simple-patient");
+        return patient;
+    }
```
Updated

c-schuler commented on Java/measureeval/src/test/java/com/lantanagroup/link/measureeval/services/PatientDataBuilder.java:
```diff
@@ -0,0 +1,34 @@
+        encounter.setSubject(new Reference("Patient/simple-patient"));
+        encounter.setStatus(Encounter.EncounterStatus.FINISHED);
+        encounter.setClass_(new Coding().setSystem("http://terminology.hl7.org/CodeSystem/v3-ActCode").setCode("AMB"));
+        return encounter;
```
Updated

c-schuler committed:
Merge branch 'dev' into me-unit-tests
